distributed/tests/test_cancelled_state.py

```diff
@@ -62,7 +75,34 @@ async def test_abort_execution_to_fetch(c, s, a, b):
     # then, a must switch the execute to fetch. Instead of doing so, it will
     # simply re-use the currently computing result.
     fut = c.submit(inc, fut, workers=[a.worker_address], key="f2")
-    await fut
+    await wait_for_state("f2", "waiting", a)
+    await ev.set()
+    assert await fut == 124  # It would be 3 if the result was copied from b
+    del fut
+    while "f1" in a.tasks:
+        await asyncio.sleep(0.01)
+
+    assert_worker_story(
+        a.story("f1"),
+        [
+            ("f1", "compute-task"),
+            ("f1", "released", "waiting", "waiting", {"f1": "ready"}),
+            ("f1", "waiting", "ready", "ready", {}),
+            ("f1", "ready", "executing", "executing", {}),
+            ("free-keys", ("f1",)),
+            ("f1", "executing", "released", "cancelled", {}),
+            ("f1", "ensure-task-exists", "cancelled"),
+            ("f1", "cancelled", "fetch", "cancelled", {"f1": ("resumed", "fetch")}),
+            ("f1", "cancelled", "resumed", "resumed", {}),
+            ("f1", "put-in-memory"),
+            ("f1", "resumed", "memory", "memory", {"f2": "ready"}),
+            ("free-keys", ("f1",)),
+            ("f1", "release-key"),
+            ("f1", "memory", "released", "released", {}),
+            ("f1", "released", "forgotten", "forgotten", {}),
+        ],
+        strict=True,
+    )
```

Comment:
I suspect that you and @fjetter are aligned here, so I'm happy to step back here.

However, my experience is that tests like these will become expensive and annoying if you ever want to change the state machine system.  Every time a dev changes the system, they will need to look at this test, understand its intent, and change it accordingly.  If this output is at all likely to change over the next few years then this seems like it might add a lot of inertia.  

Dask tests used to look a lot like this in the early days.  There was a painful process to get rid of them and replace them with tests that use the user API.

Again, I'm happy if this is the direction that you all want to go (please do not block progress on this comment).

Reply:
During the worker state machine refactoring, I stepped on a few tests that failed because they were testing a story and it changed in. Yes, it costed me an extra 10 minutes to go through them and update the story, but it was a good thing to see and validate *what* changed in the story exactly.

Reply:
Yes, this makes sense if it's one or two, and if it's the same person writing the tests as writing the system.

However, if you make dozens of these tests, then if someone follows you in a year they will either not be able to make any changes, or more likely, they will delete all of these tests.  

Anyway, I think that you and Florian have this covered.  I'm raising this as a general concern.  It may be that you aren't planning to repeat this many times, or it may be that you are planning to repeat it, but are making a reasoned decision to proceed down this path because the benefits outweight the costs.  Again, not blocking anything here.

Reply:
> However, if you make dozens of these tests, then if someone follows you in a year they will either not be able to make any changes, or more likely, they will delete all of these tests.

If somebody wants to change the state machine without understanding this, they shouldn't change the state machine at all.
Deleting these tests would be rather reckless. This has been incredibly helpful in debugging deadlocks. If these stories change, this should be done absolutely intentional and not by chance.
We should not introduce these all over the place but particularly tests that test sophisticated race conditions, this has proven to be quite useful.


One of the outcomes of the current refactoring will be the possibility to log state machine events that do not go down to this granular level. This should make tests more readable while maintaining similar expressiveness.